webui/react/src/components/UPlot/UPlotChart.tsx

```diff
@@ -26,141 +25,208 @@ interface ScaleZoomData {
+
+    // Making sure the X series and all the other series data are the same length;
+    const trimmedData = chartData.map((series) => series.slice(0, minDataLength));
+
+    // Checking to make sure the X series has some data.
+    const hasXValues = (trimmedData?.[0]?.length ?? 0) !== 0;
+
+    return [ hasXValues, trimmedData as unknown as AlignedData ];
+  }
+};
 const SCROLL_THROTTLE_TIME = 500;
 
-const UPlotChart: React.FC<Props> = ({ data, focusIndex, options, style }: Props) => {
+const UPlotChart: React.FC<Props> = ({
+  data,
+  focusIndex,
+  options,
+  style,
+  noDataMessage,
+}: Props) => {
   const chartRef = useRef<uPlot>();
   const chartDivRef = useRef<HTMLDivElement>(null);
-  const scalesRef = useRef<Record<RecordKey, uPlot.Scale>>();
   const scalesZoomData = useRef<Record<string, ScaleZoomData>>({});
-  const isZoomed = useRef<boolean>(false);
-  const mousePosition = useRef<[number, number]>();
 
-  const [ hasData, normalizedData ] = useMemo(() => {
-    if (!data || data.length < 2) return [ false, undefined ];
-
-    // Is the chart aligned (eg. linear) or faceted (eg. scatter plot)?
-    if (options?.mode === 2) {
-      return [ true, data as AlignedData ];
-    } else {
-      // Figure out the lowest sized series data.
-      const chartData = data as AlignedData;
-      const minDataLength = chartData.reduce((acc: number, series: UPlotData[]) => {
-        return Math.min(acc, series.length);
-      }, Number.MAX_SAFE_INTEGER);
+  const getAugmentedOptions = (options: Partial<uPlot.Options> | undefined): uPlot.Options =>
+    uPlot.assign(
+      {
+        hooks: {
+          setScale: [
+            (uPlot: uPlot, scaleKey: string) => {
+              // fires *after* scale change
+              if (![ 'x', 'y' ].includes(scaleKey)) return;
 
-      // Making sure the X series and all the other series data are the same length;
-      const trimmedData = chartData.map(series => series.slice(0, minDataLength));
+              // the previous max/min extent of the data
+              const prevMax = scalesZoomData.current[scaleKey]?.max;
+              const prevMin = scalesZoomData.current[scaleKey]?.min;
 
-      // Checking to make sure the X series has some data.
-      const hasXValues = trimmedData?.[0]?.length !== 0;
+              const bboxMax: number | undefined = uPlot.posToVal(
+                scaleKey === 'x' ? uPlot.bbox.width : 0,
+                scaleKey,
+              );
+              const bboxMin: number | undefined = uPlot.posToVal(
+                scaleKey === 'x' ? 0 : uPlot.bbox.height,
+                scaleKey,
+              );
 
-      return [ hasXValues, trimmedData as unknown as AlignedData ];
-    }
-  }, [ data, options?.mode ]);
+              // the new max/min extent of the data
+              const maxMax = Math.max(prevMax ?? Number.MIN_SAFE_INTEGER, bboxMax);
+              const minMin = Math.min(prevMin ?? Number.MAX_SAFE_INTEGER, bboxMin);
```

Comment:
nice, this is just to avoid having to detect `null` right?

Reply:
yes